src/P3Scheme.jl

```diff
@@ -667,4 +767,464 @@ function D_m(p3::PSP3, q::FT, N::FT, ρ_r::FT, F_r::FT) where {FT}
+ - N - number mixing ratio 
+
+ Returns the distribution of cloud particles (assumed to be of the form 
+ N'(D) = N0 * D ^ 8 * exp(-λ D ^ 3)) at given D 
+"""
+function N′cloud(D::FT, q::FT, N::FT) where {FT}
+    (λ, N_0) = get_cloud_parameters(q, N)
+    return N_0 * D^8 * exp(-λ^3 * D^3)
+end
+
+"""
+    N′_ice(D, p3, λ, N0)
+
+ - D - diameter of particle 
+ - p3 - a struct containing P3 scheme parameters
+ - λ - shape parameter of distribution 
+ - N0 - shape parameter of distribution
+
+ Returns the distribution of ice particles (assumed to be of the form 
+ N'(D) = N0 * D ^ μ * exp(-λD)) at given D 
+"""
+function N′ice(D::FT, p3::PSP3, λ::FT, N_0::FT) where {FT}
+    return N_0 * D^DSD_μ(p3, λ) * exp(-λ * D)
+end
+
+"""
+    ϕ_ice(p3, D, F_r, th) 
+
+ - p3 - a struct containing P3 Scheme parameters 
+ - D - maximum particle dimension
+ - F_r - rime mass fraction (q_rim/q_i)
+ - th - P3Scheme nonlinear solve output tuple (D_cr, D_gr, ρ_g, ρ_d)
+
+ Returns the ϕ calculation for an ice particle in the P3 Scheme with given 
+ maximum particle dimension
+"""
+function ϕ_ice(
+    D::FT,
+    p3::PSP3,
+    F_r::FT,
+    th = (; D_cr = FT(0), D_gr = FT(0), ρ_g = FT(0), ρ_d = FT(0)),
+) where {FT}
+    return 16 * p3.ρ_i^2 * (p3_area(p3, D, F_r, th))^3 /
+           (9 * π * (p3_mass(p3, D, F_r, th))^2)
+end
+
+"""
+    velocity_chen(p3, Chen2022, ρ_a, D, F_r, th)
+
+ - p3 - a struct containing P3 Scheme parameters
+ - Chen2022 - a struct with terminal velocity parameters as in Chen(2022)
+ - ρ_a - density of air  
+ - D - maximum particle dimension
+ - F_r - rime mass fraction (q_rim/q_i)
+ - th - P3Scheme nonlinear solve output tuple (D_cr, D_gr, ρ_g, ρ_d)
+
+ Returns the terminal velocity of ice at given particle dimension using P3 
+ distributions and Chen 2022 parametrizations
+"""
+function velocity_chen(
+    D::FT,
+    p3::PSP3,
+    Chen2022::CMP.Chen2022VelTypeSnowIce,
+    ρ_a::FT,
+    F_r::FT,
+    th = (; D_cr = FT(0), D_gr = FT(0), ρ_g = FT(0), ρ_d = FT(0)),
+) where {FT}
+    if D <= FT(0.000625)
+        (ai, bi, ci) = CO.Chen2022_vel_coeffs_small(Chen2022, ρ_a)
+    else
+        (ai, bi, ci) = CO.Chen2022_vel_coeffs_large(Chen2022, ρ_a)
+    end
+
+    κ = FT(-1 / 6)
+
+    v = 0
+    for i in 1:2
+        v += ai[i] * D^(bi[i]) * exp(-ci[i] * D)
+    end
+
+    return ϕ_ice(D, p3, F_r, th)^κ * v
+end
+
+"""
+    velocity_chen(p3, Chen2022, ρ_a, D)
+
+ - Chen2022 - a struct with terminal velocity parameters as in Chen(2022)
+ - ρ_a - density of air  
+ - D - maximum particle dimension
+ 
+ Returns the terminal velocity of rain given Chen 2022 velocity parametrizations
+"""
+function velocity_chen(
+    D::FT,
+    Chen2022::CMP.Chen2022VelTypeRain,
+    ρ_a::FT,
+) where {FT}
+    (ai, bi, ci) = CO.Chen2022_vel_coeffs_small(Chen2022, ρ_a)
+
+    v = 0
+    for i in 1:3
+        v += ai[i] * D^(bi[i]) * exp(-ci[i] * D)
+    end
+
+    return v
+end
+
+function get_ice_bound(p3, λ::FT, tolerance::FT) where {FT}
```

Comment:
Could you add some docstring explaining the logic how you are getting the bounds? I'm wondering if that could be also tied to the limited vars assumptions from the 2-moment scheme (or something along those lines)